File: modules/agent/src/controllers/report_controller.py
"""Controller layer for report routing to services."""

import logging
from typing import Optional, Dict, Any, List

from fastapi import Request, HTTPException

from lib.decorators import handle_http_exceptions
from serializers.report import saved_report_to_dict
from services.saved_report_service import (
    get_saved_reports,
    get_saved_report_by_id,
    create_report,
    update_report,
    delete_report,
    ReportQueryRequest,
    SavedReportCreate,
    SavedReportUpdate,
)
from schemas.report import SavedReportCreate, SavedReportUpdate
from services.report_builder import (
    execute_custom_query,
    get_lead_pipeline_report,
    get_account_overview_report,
    get_contact_coverage_report,
    get_notes_activity_report,
    get_user_audit_report,
    get_available_fields,
    generate_excel_bytes,
)

# Re-export for routes
__all__ = ["ReportQueryRequest", "SavedReportCreate", "SavedReportUpdate"]

# --- Canned Reports ---
@handle_http_exceptions
async def get_lead_pipeline(
    request: Request,
    date_from: Optional[str],
    date_to: Optional[str],
    project_id: Optional[int],
) -> Dict[str, Any]:
    """Get lead pipeline canned report."""
    tenant_id = request.state.tenant_id
    return await get_lead_pipeline_report(tenant_id, date_from, date_to, project_id)


@handle_http_exceptions
async def get_account_overview(request: Request) -> Dict[str, Any]:
    """Get account overview canned report."""
    tenant_id = request.state.tenant_id
    return await get_account_overview_report(tenant_id)


@handle_http_exceptions
async def get_contact_coverage(request: Request) -> Dict[str, Any]:
    """Get contact coverage canned report."""
    tenant_id = request.state.tenant_id
    return await get_contact_coverage_report(tenant_id)


@handle_http_exceptions
async def get_notes_activity(request: Request, project_id: Optional[int]) -> Dict[str, Any]:
    """Get notes activity canned report."""
    tenant_id = request.state.tenant_id
    return await get_notes_activity_report(tenant_id, project_id)


@handle_http_exceptions
async def get_user_audit(request: Request, user_id: Optional[int]) -> Dict[str, Any]:
    """Get user audit canned report."""
    tenant_id = request.state.tenant_id
    return await get_user_audit_report(tenant_id, user_id)


# --- Custom Report Execution ---

def get_entity_fields(entity: str) -> Dict[str, Any]:
    """Get available fields for an entity."""
    fields = get_available_fields(entity)
    if not fields["base"]:
        raise HTTPException(status_code=404, detail=f"Unknown entity: {entity}")
    return fields


@handle_http_exceptions
async def preview_custom_report(request: Request, query: ReportQueryRequest) -> Dict[str, Any]:
    """Preview a custom report with limited rows."""
    tenant_id = request.state.tenant_id
    query.limit = min(query.limit, 100)
    filters = [f.model_dump() for f in query.filters] if query.filters else None
    return await execute_custom_query(
        tenant_id=tenant_id,
        primary_entity=query.primary_entity,
        columns=query.columns,
        joins=query.joins,
        filters=filters,
        group_by=query.group_by,
        limit=query.limit,
        offset=query.offset,
        project_id=query.project_id,
    )


@handle_http_exceptions
async def run_custom_report(request: Request, query: ReportQueryRequest) -> Dict[str, Any]:
    """Run a custom report with full results."""
    tenant_id = request.state.tenant_id
    filters = [f.model_dump() for f in query.filters] if query.filters else None
    return await execute_custom_query(
        tenant_id=tenant_id,
        primary_entity=query.primary_entity,
        columns=query.columns,
        joins=query.joins,
        filters=filters,
        group_by=query.group_by,
        limit=query.limit,
        offset=query.offset,
        project_id=query.project_id,
    )


@handle_http_exceptions
async def export_custom_report(request: Request, query: ReportQueryRequest) -> tuple:
    """Export a custom report to Excel. Returns (excel_bytes, columns)."""
    tenant_id = request.state.tenant_id
    query.limit = 10000
    query.offset = 0
    filters = [f.model_dump() for f in query.filters] if query.filters else None
    result = await execute_custom_query(
        tenant_id=tenant_id,
        primary_entity=query.primary_entity,
        columns=query.columns,
        joins=query.joins,
        filters=filters,
        group_by=query.group_by,
        limit=query.limit,
        offset=query.offset,
        project_id=query.project_id,
    )
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])

    excel_bytes = generate_excel_bytes(result["data"], query.columns)
    return excel_bytes


# --- Saved Reports CRUD ---

@handle_http_exceptions
async def list_saved_reports(
    request: Request,
    project_id: Optional[int],
    include_global: bool,
    search: Optional[str],
    page: int,
    limit: int,
    sort_by: str,
    order: str,
) -> Dict[str, Any]:
    """List saved reports for the tenant with pagination and search."""
    tenant_id = request.state.tenant_id
    result = await get_saved_reports(
        tenant_id,
        project_id,
        include_global,
        search=search,
        page=page,
        limit=limit,
        sort_by=sort_by,
        sort_direction=order
    )
    return {
        "items": [saved_report_to_dict(r) for r in result["items"]],
        "total": result["total"],
        "currentPage": result["page"],
        "totalPages": result["total_pages"],
        "pageSize": result["limit"]
    }


@handle_http_exceptions
async def create_saved_report_controller(request: Request, data: SavedReportCreate) -> Dict[str, Any]:
    """Create a new saved report."""
    tenant_id = request.state.tenant_id
    user_id = getattr(request.state, "user_id", None)
    report = await create_report(
        {
            "tenant_id": tenant_id,
            "name": data.name,
            "description": data.description,
            "query_definition": data.query_definition,
            "created_by": user_id,
        },
        project_ids=data.project_ids
    )
    return saved_report_to_dict(report)


@handle_http_exceptions
async def get_saved_report(request: Request, report_id: int) -> Dict[str, Any]:
    """Get a saved report by ID."""
    tenant_id = request.state.tenant_id
    report = await get_saved_report_by_id(report_id, tenant_id)
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    return saved_report_to_dict(report)


@handle_http_exceptions
async def update_saved_report_controller(request: Request, report_id: int, data: SavedReportUpdate) -> Dict[str, Any]:
    """Update a saved report."""
    tenant_id = request.state.tenant_id
    project_ids = data.project_ids
    update_data = {
        k: v for k, v in data.model_dump().items()
        if v is not None and k != "project_ids"
    }
    if not update_data and project_ids is None:
        raise HTTPException(status_code=400, detail="No fields to update")
    report = await update_report(report_id, tenant_id, update_data, project_ids)
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    return saved_report_to_dict(report)


@handle_http_exceptions
async def delete_saved_report_controller(request: Request, report_id: int) -> Dict[str, Any]:
    """Delete a saved report."""
    tenant_id = request.state.tenant_id
    deleted = await delete_report(report_id, tenant_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Report not found")
    return {"message": "Report deleted"}

File: modules/agent/internal/controllers/report_controller.go
package controllers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"agent/internal/middleware"
	"agent/internal/schemas"
	"agent/internal/services"
)

// ReportController handles report HTTP requests
type ReportController struct {
	reportService *services.ReportService
}

// NewReportController creates a new report controller
func NewReportController(reportService *services.ReportService) *ReportController {
	return &ReportController{reportService: reportService}
}

// GetLeadPipeline handles GET /reports/canned/lead-pipeline
func (c *ReportController) GetLeadPipeline(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	dateFrom := r.URL.Query().Get("date_from")
	dateTo := r.URL.Query().Get("date_to")
	projectID := parseOptionalInt64(r.URL.Query().Get("project_id"))

	var dateFromPtr, dateToPtr *string
	if dateFrom != "" {
		dateFromPtr = &dateFrom
	}
	if dateTo != "" {
		dateToPtr = &dateTo
	}

	result, err := c.reportService.GetLeadPipelineReport(tenantID, dateFromPtr, dateToPtr, projectID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetAccountOverview handles GET /reports/canned/account-overview
func (c *ReportController) GetAccountOverview(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	result, err := c.reportService.GetAccountOverviewReport(tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetContactCoverage handles GET /reports/canned/contact-coverage
func (c *ReportController) GetContactCoverage(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	result, err := c.reportService.GetContactCoverageReport(tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetNotesActivity handles GET /reports/canned/notes-activity
func (c *ReportController) GetNotesActivity(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	projectID := parseOptionalInt64(r.URL.Query().Get("project_id"))

	result, err := c.reportService.GetNotesActivityReport(tenantID, projectID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetUserAudit handles GET /reports/canned/user-audit
func (c *ReportController) GetUserAudit(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	userID := parseOptionalInt64(r.URL.Query().Get("user_id"))

	result, err := c.reportService.GetUserAuditReport(tenantID, userID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetEntityFields handles GET /reports/fields/{entity}
func (c *ReportController) GetEntityFields(w http.ResponseWriter, r *http.Request) {
	entity := chi.URLParam(r, "entity")
	if entity == "" {
		writeError(w, http.StatusBadRequest, "entity required")
		return
	}

	result, err := c.reportService.GetEntityFields(entity)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if result == nil {
		writeError(w, http.StatusNotFound, "unknown entity: "+entity)
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// PreviewCustomReport handles POST /reports/custom/preview
func (c *ReportController) PreviewCustomReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	var query schemas.ReportQueryRequest
	if err := json.NewDecoder(r.Body).Decode(&query); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.reportService.PreviewCustomReport(tenantID, query)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// RunCustomReport handles POST /reports/custom/run
func (c *ReportController) RunCustomReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	var query schemas.ReportQueryRequest
	if err := json.NewDecoder(r.Body).Decode(&query); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.reportService.RunCustomReport(tenantID, query)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// ExportCustomReport handles POST /reports/custom/export
func (c *ReportController) ExportCustomReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	var query schemas.ReportQueryRequest
	if err := json.NewDecoder(r.Body).Decode(&query); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.reportService.ExportCustomReport(tenantID, query)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	// Return JSON data (actual Excel generation would be done client-side or with additional library)
	writeJSON(w, http.StatusOK, result)
}

// ListSavedReports handles GET /reports/saved
func (c *ReportController) ListSavedReports(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	projectID := parseOptionalInt64(r.URL.Query().Get("project_id"))
	includeGlobal := r.URL.Query().Get("include_global") != "false"
	search := r.URL.Query().Get("q")
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	sortBy := r.URL.Query().Get("sort_by")
	order := r.URL.Query().Get("order")

	if page <= 0 {
		page = 1
	}
	if limit <= 0 {
		limit = 50
	}
	if limit > 100 {
		limit = 100
	}

	result, err := c.reportService.GetSavedReports(tenantID, projectID, includeGlobal, search, page, limit, sortBy, order)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// CreateSavedReport handles POST /reports/saved
func (c *ReportController) CreateSavedReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	var userID *int64
	if uid, ok := middleware.GetUserID(r.Context()); ok {
		userID = &uid
	}

	var input schemas.SavedReportCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.reportService.CreateSavedReport(tenantID, userID, input)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, result)
}

// GetSavedReport handles GET /reports/saved/{report_id}
func (c *ReportController) GetSavedReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	reportID, err := strconv.ParseInt(chi.URLParam(r, "report_id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid report_id")
		return
	}

	result, err := c.reportService.GetSavedReport(reportID, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if result == nil {
		writeError(w, http.StatusNotFound, "report not found")
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// UpdateSavedReport handles PUT /reports/saved/{report_id}
func (c *ReportController) UpdateSavedReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	reportID, err := strconv.ParseInt(chi.URLParam(r, "report_id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid report_id")
		return
	}

	var input schemas.SavedReportUpdate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.reportService.UpdateSavedReport(reportID, tenantID, input)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if result == nil {
		writeError(w, http.StatusNotFound, "report not found")
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// DeleteSavedReport handles DELETE /reports/saved/{report_id}
func (c *ReportController) DeleteSavedReport(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.GetTenantID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "tenant required")
		return
	}

	reportID, err := strconv.ParseInt(chi.URLParam(r, "report_id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid report_id")
		return
	}

	deleted, err := c.reportService.DeleteSavedReport(reportID, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	if !deleted {
		writeError(w, http.StatusNotFound, "report not found")
		return
	}

	writeJSON(w, http.StatusOK, map[string]string{"message": "Report deleted"})
}
